lifecycle-operator/webhooks/pod_mutator/handlers/pod_annotation.go

```diff
@@ -0,0 +1,130 @@
+package handlers
+
+import (
+	"context"
+
+	argov1alpha1 "github.com/argoproj/argo-rollouts/pkg/apis/rollouts/v1alpha1"
+	"github.com/go-logr/logr"
+	apicommon "github.com/keptn/lifecycle-toolkit/lifecycle-operator/apis/lifecycle/v1alpha3/common"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+type PodAnnotationHandler struct {
+	Client client.Client
+	Log    logr.Logger
+}
+
+func (p *PodAnnotationHandler) IsAnnotated(ctx context.Context, req *admission.Request, pod *corev1.Pod) bool {
+	podIsAnnotated := isPodAnnotated(pod)
+	if !podIsAnnotated {
+		p.Log.Info("Pod is not annotated, check for parent annotations...")
+		podIsAnnotated = p.copyAnnotationsIfParentAnnotated(ctx, req, pod)
+	}
+	return podIsAnnotated
+}
+
+func (p *PodAnnotationHandler) copyAnnotationsIfParentAnnotated(ctx context.Context, req *admission.Request, pod *corev1.Pod) bool {
+	podOwner := GetOwnerReference(&pod.ObjectMeta)
+	if podOwner.UID == "" {
+		return false
+	}
+
+	switch podOwner.Kind {
+	case "ReplicaSet":
+		rs := &appsv1.ReplicaSet{}
+		if err := p.Client.Get(ctx, types.NamespacedName{Namespace: req.Namespace, Name: podOwner.Name}, rs); err != nil {
+			return false
+		}
+
+		rsOwner := GetOwnerReference(&rs.ObjectMeta)
+		if rsOwner.UID == "" {
+			return false
+		}
+
+		if rsOwner.Kind == "Rollout" {
+			ro := &argov1alpha1.Rollout{}
+			objectContainerMetaData := p.fetchParent(ctx, types.NamespacedName{Name: rsOwner.Name, Namespace: req.Namespace}, ro)
+			return copyResourceLabelsIfPresent(objectContainerMetaData, pod)
+		}
+		dp := &appsv1.Deployment{}
+		objectContainerMetaData := p.fetchParent(ctx, types.NamespacedName{Name: rsOwner.Name, Namespace: req.Namespace}, dp)
+		return copyResourceLabelsIfPresent(objectContainerMetaData, pod)
+
+	case "StatefulSet":
+		sts := &appsv1.StatefulSet{}
+		objectContainerMetaData := p.fetchParent(ctx, types.NamespacedName{Name: podOwner.Name, Namespace: req.Namespace}, sts)
+		return copyResourceLabelsIfPresent(objectContainerMetaData, pod)
+	case "DaemonSet":
+		ds := &appsv1.DaemonSet{}
+		objectContainerMetaData := p.fetchParent(ctx, types.NamespacedName{Name: podOwner.Name, Namespace: req.Namespace}, ds)
+		return copyResourceLabelsIfPresent(objectContainerMetaData, pod)
+	default:
+		return false
+	}
+}
+
+func (p *PodAnnotationHandler) fetchParent(ctx context.Context, name types.NamespacedName, objectContainer client.Object) *metav1.ObjectMeta {
+	if err := p.Client.Get(ctx, name, objectContainer); err != nil {
+		return nil
```

Comment:
shouldn't we return error here ?

Reply:
we can log it if you want, but we do not care why the parent cannot be found this could simply be because it is deleted in the meantime